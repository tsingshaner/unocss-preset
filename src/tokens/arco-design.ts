import { colorFactory, cssVar } from '@/utils/css-color-factory'
import { cssVariablesExtractor } from '@/utils/css-var-extractor'

const CSS_COLOR_ARG_TOKEN_REGEX = /^--(\w+)-\d+$/
const CSS_COLOR_FUNC_TOKEN_REGEX = /^--color-/

const cssColorArgsFilter = (v: string): boolean => Boolean(CSS_COLOR_ARG_TOKEN_REGEX.exec(v))
const cssColorFuncFilter = (v: string): boolean => Boolean(CSS_COLOR_FUNC_TOKEN_REGEX.exec(v))

/**
 * Extract Arco Design colors for unocss theme from CSS variables
 * @param code - The CSS source code
 * @returns unocss theme['colors']
 */
export const extraArcoDesignColors = (code: Uint8Array): Record<string, Record<string, string> | string> => {
  const vars = cssVariablesExtractor(code)
  const colorArgTokens = vars.filter(cssColorArgsFilter)
  const colorFunctions = vars.filter(cssColorFuncFilter)

  const color = colorFactory('rgb')

  return {
    ...Object.fromEntries(colorArgTokens.map((token) => [token.slice('--'.length), color(token)])),
    ...Object.fromEntries(colorFunctions.map((token) => [token.slice('--color-'.length), cssVar(token)]))
  }
}
