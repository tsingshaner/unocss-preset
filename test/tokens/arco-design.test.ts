import { readFile } from 'node:fs/promises'
import { resolve } from 'node:path'

import { describe } from 'vitest'

import { extraArcoDesignColors } from '@/tokens/arco-design'

describe('arco desiasync gn tokens', (test) => {
  test('snapshot', async ({ expect }) => {
    const css = await readFile(resolve(import.meta.dirname, 'fixtures/arco-design.css'), 'utf-8')
    const code = new TextEncoder().encode(css)
    const arcoTokens = extraArcoDesignColors(code)

    expect(Object.entries(arcoTokens).sort((a, b) => a[0].localeCompare(b[0]))).toMatchSnapshot()
  })
})
