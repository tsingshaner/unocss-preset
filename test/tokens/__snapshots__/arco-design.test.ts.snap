// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`arco desiasync gn tokens > snapshot 1`] = `
[
  [
    "arcoblue-1",
    "rgb(var(--arcoblue-1))",
  ],
  [
    "arcoblue-10",
    "rgb(var(--arcoblue-10))",
  ],
  [
    "arcoblue-2",
    "rgb(var(--arcoblue-2))",
  ],
  [
    "arcoblue-3",
    "rgb(var(--arcoblue-3))",
  ],
  [
    "arcoblue-4",
    "rgb(var(--arcoblue-4))",
  ],
  [
    "arcoblue-5",
    "rgb(var(--arcoblue-5))",
  ],
  [
    "arcoblue-6",
    "rgb(var(--arcoblue-6))",
  ],
  [
    "arcoblue-7",
    "rgb(var(--arcoblue-7))",
  ],
  [
    "arcoblue-8",
    "rgb(var(--arcoblue-8))",
  ],
  [
    "arcoblue-9",
    "rgb(var(--arcoblue-9))",
  ],
  [
    "bg-1",
    "var(--color-bg-1)",
  ],
  [
    "bg-2",
    "var(--color-bg-2)",
  ],
  [
    "bg-3",
    "var(--color-bg-3)",
  ],
  [
    "bg-4",
    "var(--color-bg-4)",
  ],
  [
    "bg-5",
    "var(--color-bg-5)",
  ],
  [
    "bg-popup",
    "var(--color-bg-popup)",
  ],
  [
    "bg-white",
    "var(--color-bg-white)",
  ],
  [
    "black",
    "var(--color-black)",
  ],
  [
    "blue-1",
    "rgb(var(--blue-1))",
  ],
  [
    "blue-10",
    "rgb(var(--blue-10))",
  ],
  [
    "blue-2",
    "rgb(var(--blue-2))",
  ],
  [
    "blue-3",
    "rgb(var(--blue-3))",
  ],
  [
    "blue-4",
    "rgb(var(--blue-4))",
  ],
  [
    "blue-5",
    "rgb(var(--blue-5))",
  ],
  [
    "blue-6",
    "rgb(var(--blue-6))",
  ],
  [
    "blue-7",
    "rgb(var(--blue-7))",
  ],
  [
    "blue-8",
    "rgb(var(--blue-8))",
  ],
  [
    "blue-9",
    "rgb(var(--blue-9))",
  ],
  [
    "border",
    "var(--color-border)",
  ],
  [
    "border-1",
    "var(--color-border-1)",
  ],
  [
    "border-2",
    "var(--color-border-2)",
  ],
  [
    "border-3",
    "var(--color-border-3)",
  ],
  [
    "border-4",
    "var(--color-border-4)",
  ],
  [
    "cyan-1",
    "rgb(var(--cyan-1))",
  ],
  [
    "cyan-10",
    "rgb(var(--cyan-10))",
  ],
  [
    "cyan-2",
    "rgb(var(--cyan-2))",
  ],
  [
    "cyan-3",
    "rgb(var(--cyan-3))",
  ],
  [
    "cyan-4",
    "rgb(var(--cyan-4))",
  ],
  [
    "cyan-5",
    "rgb(var(--cyan-5))",
  ],
  [
    "cyan-6",
    "rgb(var(--cyan-6))",
  ],
  [
    "cyan-7",
    "rgb(var(--cyan-7))",
  ],
  [
    "cyan-8",
    "rgb(var(--cyan-8))",
  ],
  [
    "cyan-9",
    "rgb(var(--cyan-9))",
  ],
  [
    "danger-1",
    "rgb(var(--danger-1))",
  ],
  [
    "danger-10",
    "rgb(var(--danger-10))",
  ],
  [
    "danger-2",
    "rgb(var(--danger-2))",
  ],
  [
    "danger-3",
    "rgb(var(--danger-3))",
  ],
  [
    "danger-4",
    "rgb(var(--danger-4))",
  ],
  [
    "danger-5",
    "rgb(var(--danger-5))",
  ],
  [
    "danger-6",
    "rgb(var(--danger-6))",
  ],
  [
    "danger-7",
    "rgb(var(--danger-7))",
  ],
  [
    "danger-8",
    "rgb(var(--danger-8))",
  ],
  [
    "danger-9",
    "rgb(var(--danger-9))",
  ],
  [
    "danger-light-1",
    "var(--color-danger-light-1)",
  ],
  [
    "danger-light-2",
    "var(--color-danger-light-2)",
  ],
  [
    "danger-light-3",
    "var(--color-danger-light-3)",
  ],
  [
    "danger-light-4",
    "var(--color-danger-light-4)",
  ],
  [
    "data-1",
    "var(--color-data-1)",
  ],
  [
    "data-10",
    "var(--color-data-10)",
  ],
  [
    "data-11",
    "var(--color-data-11)",
  ],
  [
    "data-12",
    "var(--color-data-12)",
  ],
  [
    "data-13",
    "var(--color-data-13)",
  ],
  [
    "data-14",
    "var(--color-data-14)",
  ],
  [
    "data-15",
    "var(--color-data-15)",
  ],
  [
    "data-16",
    "var(--color-data-16)",
  ],
  [
    "data-17",
    "var(--color-data-17)",
  ],
  [
    "data-18",
    "var(--color-data-18)",
  ],
  [
    "data-19",
    "var(--color-data-19)",
  ],
  [
    "data-2",
    "var(--color-data-2)",
  ],
  [
    "data-20",
    "var(--color-data-20)",
  ],
  [
    "data-3",
    "var(--color-data-3)",
  ],
  [
    "data-4",
    "var(--color-data-4)",
  ],
  [
    "data-5",
    "var(--color-data-5)",
  ],
  [
    "data-6",
    "var(--color-data-6)",
  ],
  [
    "data-7",
    "var(--color-data-7)",
  ],
  [
    "data-8",
    "var(--color-data-8)",
  ],
  [
    "data-9",
    "var(--color-data-9)",
  ],
  [
    "fill-1",
    "var(--color-fill-1)",
  ],
  [
    "fill-2",
    "var(--color-fill-2)",
  ],
  [
    "fill-3",
    "var(--color-fill-3)",
  ],
  [
    "fill-4",
    "var(--color-fill-4)",
  ],
  [
    "gold-1",
    "rgb(var(--gold-1))",
  ],
  [
    "gold-10",
    "rgb(var(--gold-10))",
  ],
  [
    "gold-2",
    "rgb(var(--gold-2))",
  ],
  [
    "gold-3",
    "rgb(var(--gold-3))",
  ],
  [
    "gold-4",
    "rgb(var(--gold-4))",
  ],
  [
    "gold-5",
    "rgb(var(--gold-5))",
  ],
  [
    "gold-6",
    "rgb(var(--gold-6))",
  ],
  [
    "gold-7",
    "rgb(var(--gold-7))",
  ],
  [
    "gold-8",
    "rgb(var(--gold-8))",
  ],
  [
    "gold-9",
    "rgb(var(--gold-9))",
  ],
  [
    "gray-1",
    "rgb(var(--gray-1))",
  ],
  [
    "gray-10",
    "rgb(var(--gray-10))",
  ],
  [
    "gray-2",
    "rgb(var(--gray-2))",
  ],
  [
    "gray-3",
    "rgb(var(--gray-3))",
  ],
  [
    "gray-4",
    "rgb(var(--gray-4))",
  ],
  [
    "gray-5",
    "rgb(var(--gray-5))",
  ],
  [
    "gray-6",
    "rgb(var(--gray-6))",
  ],
  [
    "gray-7",
    "rgb(var(--gray-7))",
  ],
  [
    "gray-8",
    "rgb(var(--gray-8))",
  ],
  [
    "gray-9",
    "rgb(var(--gray-9))",
  ],
  [
    "green-1",
    "rgb(var(--green-1))",
  ],
  [
    "green-10",
    "rgb(var(--green-10))",
  ],
  [
    "green-2",
    "rgb(var(--green-2))",
  ],
  [
    "green-3",
    "rgb(var(--green-3))",
  ],
  [
    "green-4",
    "rgb(var(--green-4))",
  ],
  [
    "green-5",
    "rgb(var(--green-5))",
  ],
  [
    "green-6",
    "rgb(var(--green-6))",
  ],
  [
    "green-7",
    "rgb(var(--green-7))",
  ],
  [
    "green-8",
    "rgb(var(--green-8))",
  ],
  [
    "green-9",
    "rgb(var(--green-9))",
  ],
  [
    "lime-1",
    "rgb(var(--lime-1))",
  ],
  [
    "lime-10",
    "rgb(var(--lime-10))",
  ],
  [
    "lime-2",
    "rgb(var(--lime-2))",
  ],
  [
    "lime-3",
    "rgb(var(--lime-3))",
  ],
  [
    "lime-4",
    "rgb(var(--lime-4))",
  ],
  [
    "lime-5",
    "rgb(var(--lime-5))",
  ],
  [
    "lime-6",
    "rgb(var(--lime-6))",
  ],
  [
    "lime-7",
    "rgb(var(--lime-7))",
  ],
  [
    "lime-8",
    "rgb(var(--lime-8))",
  ],
  [
    "lime-9",
    "rgb(var(--lime-9))",
  ],
  [
    "link-1",
    "rgb(var(--link-1))",
  ],
  [
    "link-10",
    "rgb(var(--link-10))",
  ],
  [
    "link-2",
    "rgb(var(--link-2))",
  ],
  [
    "link-3",
    "rgb(var(--link-3))",
  ],
  [
    "link-4",
    "rgb(var(--link-4))",
  ],
  [
    "link-5",
    "rgb(var(--link-5))",
  ],
  [
    "link-6",
    "rgb(var(--link-6))",
  ],
  [
    "link-7",
    "rgb(var(--link-7))",
  ],
  [
    "link-8",
    "rgb(var(--link-8))",
  ],
  [
    "link-9",
    "rgb(var(--link-9))",
  ],
  [
    "link-light-1",
    "var(--color-link-light-1)",
  ],
  [
    "link-light-2",
    "var(--color-link-light-2)",
  ],
  [
    "link-light-3",
    "var(--color-link-light-3)",
  ],
  [
    "link-light-4",
    "var(--color-link-light-4)",
  ],
  [
    "magenta-1",
    "rgb(var(--magenta-1))",
  ],
  [
    "magenta-10",
    "rgb(var(--magenta-10))",
  ],
  [
    "magenta-2",
    "rgb(var(--magenta-2))",
  ],
  [
    "magenta-3",
    "rgb(var(--magenta-3))",
  ],
  [
    "magenta-4",
    "rgb(var(--magenta-4))",
  ],
  [
    "magenta-5",
    "rgb(var(--magenta-5))",
  ],
  [
    "magenta-6",
    "rgb(var(--magenta-6))",
  ],
  [
    "magenta-7",
    "rgb(var(--magenta-7))",
  ],
  [
    "magenta-8",
    "rgb(var(--magenta-8))",
  ],
  [
    "magenta-9",
    "rgb(var(--magenta-9))",
  ],
  [
    "mask-bg",
    "var(--color-mask-bg)",
  ],
  [
    "menu-dark-bg",
    "var(--color-menu-dark-bg)",
  ],
  [
    "menu-dark-hover",
    "var(--color-menu-dark-hover)",
  ],
  [
    "menu-light-bg",
    "var(--color-menu-light-bg)",
  ],
  [
    "neutral-1",
    "var(--color-neutral-1)",
  ],
  [
    "neutral-10",
    "var(--color-neutral-10)",
  ],
  [
    "neutral-2",
    "var(--color-neutral-2)",
  ],
  [
    "neutral-3",
    "var(--color-neutral-3)",
  ],
  [
    "neutral-4",
    "var(--color-neutral-4)",
  ],
  [
    "neutral-5",
    "var(--color-neutral-5)",
  ],
  [
    "neutral-6",
    "var(--color-neutral-6)",
  ],
  [
    "neutral-7",
    "var(--color-neutral-7)",
  ],
  [
    "neutral-8",
    "var(--color-neutral-8)",
  ],
  [
    "neutral-9",
    "var(--color-neutral-9)",
  ],
  [
    "orange-1",
    "rgb(var(--orange-1))",
  ],
  [
    "orange-10",
    "rgb(var(--orange-10))",
  ],
  [
    "orange-2",
    "rgb(var(--orange-2))",
  ],
  [
    "orange-3",
    "rgb(var(--orange-3))",
  ],
  [
    "orange-4",
    "rgb(var(--orange-4))",
  ],
  [
    "orange-5",
    "rgb(var(--orange-5))",
  ],
  [
    "orange-6",
    "rgb(var(--orange-6))",
  ],
  [
    "orange-7",
    "rgb(var(--orange-7))",
  ],
  [
    "orange-8",
    "rgb(var(--orange-8))",
  ],
  [
    "orange-9",
    "rgb(var(--orange-9))",
  ],
  [
    "orangered-1",
    "rgb(var(--orangered-1))",
  ],
  [
    "orangered-10",
    "rgb(var(--orangered-10))",
  ],
  [
    "orangered-2",
    "rgb(var(--orangered-2))",
  ],
  [
    "orangered-3",
    "rgb(var(--orangered-3))",
  ],
  [
    "orangered-4",
    "rgb(var(--orangered-4))",
  ],
  [
    "orangered-5",
    "rgb(var(--orangered-5))",
  ],
  [
    "orangered-6",
    "rgb(var(--orangered-6))",
  ],
  [
    "orangered-7",
    "rgb(var(--orangered-7))",
  ],
  [
    "orangered-8",
    "rgb(var(--orangered-8))",
  ],
  [
    "orangered-9",
    "rgb(var(--orangered-9))",
  ],
  [
    "pinkpurple-1",
    "rgb(var(--pinkpurple-1))",
  ],
  [
    "pinkpurple-10",
    "rgb(var(--pinkpurple-10))",
  ],
  [
    "pinkpurple-2",
    "rgb(var(--pinkpurple-2))",
  ],
  [
    "pinkpurple-3",
    "rgb(var(--pinkpurple-3))",
  ],
  [
    "pinkpurple-4",
    "rgb(var(--pinkpurple-4))",
  ],
  [
    "pinkpurple-5",
    "rgb(var(--pinkpurple-5))",
  ],
  [
    "pinkpurple-6",
    "rgb(var(--pinkpurple-6))",
  ],
  [
    "pinkpurple-7",
    "rgb(var(--pinkpurple-7))",
  ],
  [
    "pinkpurple-8",
    "rgb(var(--pinkpurple-8))",
  ],
  [
    "pinkpurple-9",
    "rgb(var(--pinkpurple-9))",
  ],
  [
    "primary-1",
    "rgb(var(--primary-1))",
  ],
  [
    "primary-10",
    "rgb(var(--primary-10))",
  ],
  [
    "primary-2",
    "rgb(var(--primary-2))",
  ],
  [
    "primary-3",
    "rgb(var(--primary-3))",
  ],
  [
    "primary-4",
    "rgb(var(--primary-4))",
  ],
  [
    "primary-5",
    "rgb(var(--primary-5))",
  ],
  [
    "primary-6",
    "rgb(var(--primary-6))",
  ],
  [
    "primary-7",
    "rgb(var(--primary-7))",
  ],
  [
    "primary-8",
    "rgb(var(--primary-8))",
  ],
  [
    "primary-9",
    "rgb(var(--primary-9))",
  ],
  [
    "primary-light-1",
    "var(--color-primary-light-1)",
  ],
  [
    "primary-light-2",
    "var(--color-primary-light-2)",
  ],
  [
    "primary-light-3",
    "var(--color-primary-light-3)",
  ],
  [
    "primary-light-4",
    "var(--color-primary-light-4)",
  ],
  [
    "purple-1",
    "rgb(var(--purple-1))",
  ],
  [
    "purple-10",
    "rgb(var(--purple-10))",
  ],
  [
    "purple-2",
    "rgb(var(--purple-2))",
  ],
  [
    "purple-3",
    "rgb(var(--purple-3))",
  ],
  [
    "purple-4",
    "rgb(var(--purple-4))",
  ],
  [
    "purple-5",
    "rgb(var(--purple-5))",
  ],
  [
    "purple-6",
    "rgb(var(--purple-6))",
  ],
  [
    "purple-7",
    "rgb(var(--purple-7))",
  ],
  [
    "purple-8",
    "rgb(var(--purple-8))",
  ],
  [
    "purple-9",
    "rgb(var(--purple-9))",
  ],
  [
    "red-1",
    "rgb(var(--red-1))",
  ],
  [
    "red-10",
    "rgb(var(--red-10))",
  ],
  [
    "red-2",
    "rgb(var(--red-2))",
  ],
  [
    "red-3",
    "rgb(var(--red-3))",
  ],
  [
    "red-4",
    "rgb(var(--red-4))",
  ],
  [
    "red-5",
    "rgb(var(--red-5))",
  ],
  [
    "red-6",
    "rgb(var(--red-6))",
  ],
  [
    "red-7",
    "rgb(var(--red-7))",
  ],
  [
    "red-8",
    "rgb(var(--red-8))",
  ],
  [
    "red-9",
    "rgb(var(--red-9))",
  ],
  [
    "secondary",
    "var(--color-secondary)",
  ],
  [
    "secondary-active",
    "var(--color-secondary-active)",
  ],
  [
    "secondary-disabled",
    "var(--color-secondary-disabled)",
  ],
  [
    "secondary-hover",
    "var(--color-secondary-hover)",
  ],
  [
    "spin-layer-bg",
    "var(--color-spin-layer-bg)",
  ],
  [
    "success-1",
    "rgb(var(--success-1))",
  ],
  [
    "success-10",
    "rgb(var(--success-10))",
  ],
  [
    "success-2",
    "rgb(var(--success-2))",
  ],
  [
    "success-3",
    "rgb(var(--success-3))",
  ],
  [
    "success-4",
    "rgb(var(--success-4))",
  ],
  [
    "success-5",
    "rgb(var(--success-5))",
  ],
  [
    "success-6",
    "rgb(var(--success-6))",
  ],
  [
    "success-7",
    "rgb(var(--success-7))",
  ],
  [
    "success-8",
    "rgb(var(--success-8))",
  ],
  [
    "success-9",
    "rgb(var(--success-9))",
  ],
  [
    "success-light-1",
    "var(--color-success-light-1)",
  ],
  [
    "success-light-2",
    "var(--color-success-light-2)",
  ],
  [
    "success-light-3",
    "var(--color-success-light-3)",
  ],
  [
    "success-light-4",
    "var(--color-success-light-4)",
  ],
  [
    "text-1",
    "var(--color-text-1)",
  ],
  [
    "text-2",
    "var(--color-text-2)",
  ],
  [
    "text-3",
    "var(--color-text-3)",
  ],
  [
    "text-4",
    "var(--color-text-4)",
  ],
  [
    "tooltip-bg",
    "var(--color-tooltip-bg)",
  ],
  [
    "warning-1",
    "rgb(var(--warning-1))",
  ],
  [
    "warning-10",
    "rgb(var(--warning-10))",
  ],
  [
    "warning-2",
    "rgb(var(--warning-2))",
  ],
  [
    "warning-3",
    "rgb(var(--warning-3))",
  ],
  [
    "warning-4",
    "rgb(var(--warning-4))",
  ],
  [
    "warning-5",
    "rgb(var(--warning-5))",
  ],
  [
    "warning-6",
    "rgb(var(--warning-6))",
  ],
  [
    "warning-7",
    "rgb(var(--warning-7))",
  ],
  [
    "warning-8",
    "rgb(var(--warning-8))",
  ],
  [
    "warning-9",
    "rgb(var(--warning-9))",
  ],
  [
    "warning-light-1",
    "var(--color-warning-light-1)",
  ],
  [
    "warning-light-2",
    "var(--color-warning-light-2)",
  ],
  [
    "warning-light-3",
    "var(--color-warning-light-3)",
  ],
  [
    "warning-light-4",
    "var(--color-warning-light-4)",
  ],
  [
    "white",
    "var(--color-white)",
  ],
  [
    "yellow-1",
    "rgb(var(--yellow-1))",
  ],
  [
    "yellow-10",
    "rgb(var(--yellow-10))",
  ],
  [
    "yellow-2",
    "rgb(var(--yellow-2))",
  ],
  [
    "yellow-3",
    "rgb(var(--yellow-3))",
  ],
  [
    "yellow-4",
    "rgb(var(--yellow-4))",
  ],
  [
    "yellow-5",
    "rgb(var(--yellow-5))",
  ],
  [
    "yellow-6",
    "rgb(var(--yellow-6))",
  ],
  [
    "yellow-7",
    "rgb(var(--yellow-7))",
  ],
  [
    "yellow-8",
    "rgb(var(--yellow-8))",
  ],
  [
    "yellow-9",
    "rgb(var(--yellow-9))",
  ],
]
`;
